index.js

```diff
@@ -125,6 +136,460 @@
+      }
+    }
+    return upcarrots + '\n' + highlightNumbers;
+  };
+
+  //  sigGetName :: Node -> String
+  var sigGetName = function(sig) {
+    return sig.children[SIG_IDX_NAME].children[0].text;
+  };
+
+  //  findConstraint :: (TypeClass, String) -> Node -> Nullable Node
+  var findConstraint = function(typeClass, typeVarName) {
+    return function(sig) {
+      return traverseFind(function(node, parents) {
+        if (isVar(node) && node.text === typeVarName) {
+          var parent = getParent(parents);
+          if (parent.text === getConstraintText(typeClass)) {
+            return parent;
+          }
+        }
+      }, sig.children[SIG_IDX_CONSTRAINTS]);
+    };
+  };
+
+  //  findTypeVar :: (String, Node) -> Nullable Node
+  var findTypeVar = function(typeVarName, startNode) {
+    return traverseFind(function(node, parents) {
+      if (isVar(node) && node.text === typeVarName) {
+        return node;
+      }
+    }, startNode);
+  };
+
+  //  findArg :: (Integer, Node) -> Node
+  var findArg = function(index, sig) {
+    var args = nonDelimiters(sig.children[SIG_IDX_ARGS].children);
+    return args[index];
+  };
+
+  //  findArgTypeVar :: (Integer, String) -> Node -> Nullable Node
+  var findArgTypeVar = function(index, typeVarName) {
+    return function(sig) {
+      return findTypeVar(typeVarName, findArg(index, sig));
+    };
+  };
+
+  //  findVal :: Integer -> Node -> Node
+  var findVal = function(index) {
+    return function(sig) {
+      return isNaN(index) ? sig.children[SIG_IDX_RET] :
+                            findArg(index, sig);
+    };
+  };
+
+  //  findRetTypeVar :: String -> Node -> Node
+  var findRetTypeVar = function(typeVarName) {
+    return function(sig) {
+      return findTypeVar(typeVarName, sig.children[SIG_IDX_RET]);
+    };
+  };
+
+  //  pipe :: (Any -> [(Any -> Any)]) -> Any
+  var pipe = function(data, fns) {
+    var r = fns[0](data);
+    for (var idx = 1; idx < fns.length; idx += 1) {
+      r = fns[idx](r);
+    }
+    return r;
+  };
+
+  //  head :: [a] -> a?
+  var head = function(xs) {
+    return xs[0];
+  };
+
+  //  tail :: [a] -> [a]
+  var tail = function(xs) {
+    return xs.slice(1);
+  };
+
+  //  last :: [a] -> a?
+  var last = function(xs) {
+    return xs[xs.length - 1];
+  };
+
+  //  toPairs :: StrMap a -> [Pair String a]
+  var toPairs = function(obj) {
+    var pairs = [];
+    for (var k in obj) {
+      pairs.push([k, obj[k]]);
+    }
+    return pairs;
+  };
+
+  //  isFirstChild :: (Node, Node) -> Boolean
+  var isFirstChild = function(node, parent) {
+    return head(nonDelimiters(parent.children)) === node;
+  };
+
+  //  isRootFirstChild :: (Node, [Node]) -> Boolean
+  var isRootFirstChild = function(node, parents) {
+    return (
+      isRootChild(parents) &&
+      isFirstChild(node, head(parents))
+    );
+  };
+
+  //  getParent :: [Node] -> Nullable Node
+  var getParent = last;
+
+  //  isRoot :: [Node] -> Boolean
+  var isRoot = isEmpty;
```

Comment:
> Could we instead move determineActualTypes out of create? It needs access to the environment, so that would need to be provided as an argument. This is not too much of an inconvenience if one makes it a curried function:

I considered this a while ago for another issue, but it won't help for my current problem. Previously, types had been determined and passed as arguments to the error functions. This always bothered me a little but needed to be done because of where `determineActualTypes` was. Now it is not so simple as passing the types since we display things differently for collections. To solve this I pass the `valueTypes` function (which itself calls `determineActualTypes`) as an argument. So I would instead either need to pass around a curried function or `env` itself. Seems a bit awkward to always be passing this same function to all of our error functions. But it works, so not a big deal if you like the error functions where they are.